File: src/main/java/org/myfintech/payment/entity/AbstractEntity.java
/*
 * Copyright (c) 2024 MyFintech Payment Service
 * All rights reserved.
 *
 * This software is proprietary and confidential. Unauthorized copying of this file,
 * via any medium, is strictly prohibited.
 * @author : Dhanuka Ranasinghe
 * @since : Date: 11/07/2025
 */
package org.myfintech.payment.entity;

import java.time.OffsetDateTime;

import jakarta.persistence.Column;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.MappedSuperclass;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import lombok.AllArgsConstructor;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import static org.myfintech.payment.entity.EntityColumnConstants.*;

@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@MappedSuperclass
@EqualsAndHashCode(onlyExplicitlyIncluded = true)
public abstract class AbstractEntity<T> {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @EqualsAndHashCode.Include
    protected T id;

    @Column(name = COMMON_CREATED_DATE, nullable = false, updatable = false)
    private OffsetDateTime createdDate;

    @Column(name = COMMON_MODIFIED_DATE, nullable = false)
    private OffsetDateTime modifiedDate;

    @PrePersist
    protected void onCreate() {
        OffsetDateTime now = OffsetDateTime.now();
        this.createdDate = now;
        this.modifiedDate = now;
    }

    @PreUpdate
    protected void onUpdate() {
        this.modifiedDate = OffsetDateTime.now();
    }
}

File: src/main/kotlin/org/myfintech/payment/entity/AbstractEntity.kt
package org.myfintech.payment.entity

import jakarta.persistence.*
import java.time.OffsetDateTime
import org.hibernate.Hibernate

@MappedSuperclass
open class AbstractEntity<T : Any>(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = EntityColumnConstants.COMMON_ID, updatable = false, nullable = false)
    open var id: T? = null,

    @Column(name = EntityColumnConstants.COMMON_CREATED_DATE, nullable = false, updatable = false)
    open var createdDate: OffsetDateTime? = null,

    @Column(name = EntityColumnConstants.COMMON_MODIFIED_DATE, nullable = false)
    open var modifiedDate: OffsetDateTime? = null
) {
    @PrePersist
    protected fun onCreate() {
        val now = OffsetDateTime.now()
        createdDate = now
        modifiedDate = now
    }

    @PreUpdate
    protected fun onUpdate() {
        modifiedDate = OffsetDateTime.now()
    }

    // === equals/hashCode equivalent to Lombok @EqualsAndHashCode(onlyExplicitlyIncluded = true) on id ===
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null) return false
        // proxy-safe type check
        if (Hibernate.getClass(this) != Hibernate.getClass(other)) return false
        other as AbstractEntity<*>
        // only id participates in equality
        return id != null && id == other.id
    }

    override fun hashCode(): Int = Hibernate.getClass(this).hashCode()
}
